### DISCUSSION TITLE
Steps for data-parallel DeepXDE

### DISCUSSION THREAD

**Discussion by pescap** [lululxvi/deepxde#1277] [Ideas]
- [x] DeepXDE wak scaling for `tensorflow.compat.v1`: https://github.com/lululxvi/deepxde/pull/1205
- [x] Update Docker image with `mpi4py`: https://github.com/lululxvi/deepxde/pull/1278
- [x] Add parameter for scaling mode (weak or strong): https://github.com/lululxvi/deepxde/pull/1279
- [x] Separate the weak and strong scaling implementations: https://github.com/lululxvi/deepxde/pull/1282
- [x] Define custom parameters for `hvd.DsitributedOptimizer`: https://github.com/lululxvi/deepxde/pull/1285
- [x] Implement strong scaling for `tensorflow.compat.v1`: https://github.com/lululxvi/deepxde/pull/1284
- [ ] Add documentation for Data Parallel DeepXDE
- [ ] Study `Sobol` distribution for weak scaling
- [ ] Study resampling strategies for weak scaling

**Comment by pescap** [comment]
Concerning documentation for data parallel DeepXDE, I would create a section after Demos of Operator Learning and before FAQ, do you agree @lululxvi?

The name would be: Data parallel acceleration.

**Reply by lululxvi** [reply]
Sounds good.

**Comment by pescap** [comment]
Hi! Now, `pde.PDE` is with data parallel sampling.

According to you @lululxvi, are there other `class(Data)` that shall be accelerated?  

Also, it would be interesting to study  the data parallel acceleration for DeepONets. Which data class or example do you suggest me to start with?

Thank you!

**Reply by lululxvi** [reply]
Maybe double check `pde.TimePDE`.

Yes, DeepONet parallel training is a good next step. You can simply use the demo examples.

**Reply by pescap** [reply]
Hi, how can I obtain the training points in:

https://github.com/lululxvi/deepxde/blob/4363618db90375de7d1b0b2c9b3ee9190730c225/examples/operator/antiderivative_aligned.py#L7

Was `antiderivative_aligned_train.npz` too large to be added to the repository?

**Reply by pescap** [reply]
Sorry I did not read the [documentation](https://deepxde.readthedocs.io/en/latest/demos/operator/antiderivative_aligned.html).

As stated in the tutorial, the datasets are available [here](https://drive.google.com/drive/folders/1Sb0cLnMv8rCeQyBUk97VyKa5Le9BKQNb).

A few comments:
- Why not adding directly the datasets to DeepXDE?
- It would be interesting to release the code to generate the datasets.

**Reply by lululxvi** [reply]
- I want to keep the repo size small.
- It is a very simple ODE. The code can be found at https://github.com/lululxvi/deeponet

**Comment by pescap** [comment]
Hi @lululxvi, it would be very useful to add the capability for:

https://github.com/lululxvi/deepxde/blob/47127ac60496a52ccd01d17fcea9b67cc3afc5ad/deepxde/optimizers/tensorflow_compat_v1/optimizers.py#L59

to have other parameters. For example, Horovod has Adasum algorithm, or fp16 compression. See:

https://github.com/pescap/horovodPINNs/blob/46aa5faee24f9512d2a9e7ae136286c558d9f94b/Laplace/hvd_hval.py#LL155C1-L161C1

Are you ok if I follow a similar structure to what was used for L-BFGS (defining the parameters as an external dictionary)?

**Reply by lululxvi** [reply]
Good idea.

**Comment by pescap** [comment]
Hi @lululxvi, which is the easiest way to obtain the training domain points (to split them appropriately)?

Wouldn't it be easier to define here: https://github.com/lululxvi/deepxde/blob/742ef4dba7e5ec7e3284112bc490b5f2494dce3f/deepxde/data/pde.py#L263

```
self.train_domain = X
```

When I apply strong scaling, I want to split the domain training points over ranks. Here:

https://github.com/lululxvi/deepxde/blob/742ef4dba7e5ec7e3284112bc490b5f2494dce3f/deepxde/data/pde.py#L180

it sounds like they correspond to the `num_domain` last rows. Finally, I also have to take the anchor points into account.

**Reply by pescap** [reply]
I am a little at loss with the different bc training point definitions:


https://github.com/lululxvi/deepxde/blob/742ef4dba7e5ec7e3284112bc490b5f2494dce3f/deepxde/data/pde.py#L264C1-L270


and 

https://github.com/lululxvi/deepxde/blob/742ef4dba7e5ec7e3284112bc490b5f2494dce3f/deepxde/data/pde.py#L284-L292

**Reply by lululxvi** [reply]
Do the PDE points splitting for `self.train_x_all` at 
https://github.com/lululxvi/deepxde/blob/742ef4dba7e5ec7e3284112bc490b5f2494dce3f/deepxde/data/pde.py#L194

**Reply by pescap** [reply]
Is it the same as `train_x_all` as in:

https://github.com/lululxvi/deepxde/blob/742ef4dba7e5ec7e3284112bc490b5f2494dce3f/deepxde/data/pde.py#L180

**Reply by lululxvi** [reply]
Yes, but we need to do the splitting after computing the BC points.

**Reply by pescap** [reply]
Ok, I understand.

From what I observe, `self.train_x_all` has `num_domain + num_boundary` points. I am not sure to understand why. I assumed that the PDE points were defined in the interior of the domain.

**Reply by lululxvi** [reply]
Boundary points are also used for PDE loss.

**Reply by pescap** [reply]
According to me, the PDE is not valid on the boundary (only in the interior of the domain), as the operators are defined over open domains. Could you please give more details about this choice?

**Reply by lululxvi** [reply]
In general, boundary points should also satisfy the PDE. Could you give a counterexample?
